py-polars/polars/_cpu_check.py

```diff
@@ -1,13 +1,249 @@
+        self.func_ptr = func_type(self.addr)
+
+    def __call__(self, eax: int, ecx: int = 0) -> CPUID_struct:
+        struct = CPUID_struct()
+        self.func_ptr(struct, eax, ecx)
+        return struct
+
+    def __del__(self) -> None:
+        if _IS_WINDOWS:
+            self.win.VirtualFree.restype = c_long
+            self.win.VirtualFree.argtypes = [c_void_p, c_size_t, c_ulong]
+            self.win.VirtualFree(self.addr, 0, _MEM_RELEASE)
+
+
+def read_cpu_flags() -> dict[str, bool]:
+    # Right now we only enable extra feature flags for x86.
+    if _POLARS_ARCH != "x86-64":
+        return {}
+
+    # CPU flags from https://en.wikipedia.org/wiki/CPUID
+    cpuid = CPUID()
+    cpuid1 = cpuid(1, 0)
+    cpuid7 = cpuid(7, 0)
+    cpuid81h = cpuid(0x80000001, 0)
+
+    return {
+        "sse3": bool(cpuid1.ecx & (1 << 0)),
+        "ssse3": bool(cpuid1.ecx & (1 << 9)),
+        "fma": bool(cpuid1.ecx & (1 << 12)),
+        "sse4.1": bool(cpuid1.ecx & (1 << 19)),
+        "sse4.2": bool(cpuid1.ecx & (1 << 20)),
+        "popcnt": bool(cpuid1.ecx & (1 << 23)),
+        "avx": bool(cpuid1.ecx & (1 << 28)),
+        "bmi1": bool(cpuid7.ebx & (1 << 3)),
+        "bmi2": bool(cpuid7.ebx & (1 << 8)),
+        "avx2": bool(cpuid7.ebx & (1 << 5)),
+        "lzcnt": bool(cpuid81h.ecx & (1 << 5)),
+    }
+
+
+def check_cpu_flags() -> None:
+    if (
+        not _POLARS_FEATURE_FLAGS
+        or _POLARS_LTS_CPU
+        or os.environ.get("POLARS_SKIP_CPU_CHECK")
+    ):
+        return
+
+    expected_cpu_flags = [f.lstrip("+") for f in _POLARS_FEATURE_FLAGS.split(",")]
+    supported_cpu_flags = read_cpu_flags()
+
+    missing_features = []
+    for f in expected_cpu_flags:
+        if f not in supported_cpu_flags:
+            raise RuntimeError(f'unknown feature flag "{f}"')
+
+        if not supported_cpu_flags[f]:
+            missing_features.append(f)
+
+    if missing_features:
+        import warnings  # Only import if necessary.
+
+        warnings.warn(
+            f"""Missing required CPU features.
+
+This version of Polars requires the following CPU features your processor appears to be missing:
+    {", ".join(missing_features)}
+Continuing to use this version of Polars on this processor will likely result in a crash. If you believe this to be a false positive, please set the `POLARS_SKIP_CPU_CHECK` environment variable to bypass this check.
+
+If you are on an Apple ARM machine (e.g. M1) this is likely due to running Python under Rosetta. It is recommended to install a native version of Python that does not run under Rosetta x86-64 emulation.
 
-# TODO: Implement this.
-def __verify_cpu_support() -> None:
-    """Raise an informative error if Polars cannot be run on the current CPU."""
-    ...
+Otherwise, please try the `polars-lts-cpu` package instead of `polars`. This package is compiled with fewer CPU flags to support older / less featureful processors.
```

Comment:
I would change the ordering / wording somewhat. I think it should be a bit less wordy and emphasize installing `polars-lts-cpu` rather than first outlining the Apple case / false positive message.

```
Missing required CPU features.

The following CPU features were not detected:
...

Continuing to use this version of Polars on this processor will likely result in a crash.
Install the `polars-lts-cpu` package instead of `polars` to run Polars with better compatibility.

Hint: If you are on an Apple ARM machine (e.g. M1) this is likely due to running Python under Rosetta.
It is recommended to install a native version of Python that does not run under Rosetta x86-64 emulation.

If you believe this warning to be a false positive, you can set the `POLARS_SKIP_CPU_CHECK` environment variable to bypass this check.

```

Up to you though if you think this makes sense. The existing message is already OK.